src/awkward/_connect/numba/layoutbuilder.py

```diff
@@ -0,0 +1,2547 @@
+            val._valid_when,
+            val._lsb_order,
+            val._parameters,
+        )
+
+    elif isinstance(val, ByteMasked):
+        return ByteMaskedType(
+            numba.from_dtype(val._mask.dtype),
+            val._content,
+            val._valid_when,
+            val._parameters,
+        )
+
+    elif isinstance(val, Empty):
+        return EmptyType(val._parameters)
+
+    elif isinstance(val, IndexedOption):
+        return IndexedOptionType(
+            numba.from_dtype(val._index.dtype), val._content, val._parameters
+        )
+
+    elif isinstance(val, ListOffset):
+        return ListOffsetType(
+            numba.from_dtype(val._offsets.dtype), val._content, val._parameters
+        )
+
+    elif isinstance(val, Numpy):
+        return NumpyType(numba.from_dtype(val._data.dtype), val._parameters)
+
+    elif isinstance(val, Record):
+        return RecordType(val._contents, val._fields, val._parameters)
+
+    elif isinstance(val, Regular):
+        return RegularType(val._content, val._size, val._parameters)
+
+    elif isinstance(val, Tuple):
+        return TupleType(val._contents, val._parameters)
+
+    elif isinstance(val, Union):
+        return UnionType(
+            numba.from_dtype(val._tags.dtype),
+            numba.from_dtype(val._index.dtype),
+            val._contents,
+            val._parameters,
+        )
+
+    elif isinstance(val, Unmasked):
+        return UnmaskedType(val._content, val._parameters)
+
+    else:
+        raise TypeError("unrecognized LayoutBuilder type")
+
+
+@numba.extending.overload(len)
+def LayoutBuilderType_len(builder):
+    if isinstance(
+        builder,
+        (
+            BitMaskedType,
+            ByteMaskedType,
+            EmptyType,
+            IndexedOptionType,
+            ListOffsetType,
+            NumpyType,
+            RecordType,
+            RegularType,
+            TupleType,
+            UnionType,
+            UnmaskedType,
+        ),
+    ):
+
+        def len_impl(builder):
+            return builder._length_get()
+
+        return len_impl
+
+
+########## Numpy ############################################################
+
+
+@final
+class Numpy(LayoutBuilder):
+    def __init__(self, dtype, *, parameters=None, initial=1024, resize=8.0):
+        self._data = GrowableBuffer(dtype=dtype, initial=initial, resize=resize)
+        self._parameters = parameters
+
+    @classmethod
+    def _from_buffer(cls, data):
+        out = cls.__new__(cls)
+        out._data = data
+        out._parameters = None
+        return out
+
+    def __repr__(self):
+        return f"ak.numba.lb.Numpy({self._data.dtype}, parameters={self._parameters})"
+
+    def numbatype(self):
+        return NumpyType(
+            numba.from_dtype(self.dtype), numba.types.StringLiteral(self._parameters)
+        )
+
+    @property
+    def _length(self):
+        return len(self._data)
+
+    def __len__(self):
+        return self._length
```

Comment:
```suggestion
    def __len__(self):
        return len(self._data)
```

and similarly for all other `LayoutBuilder` subclasses. The `Content` subclasses have extra properties for lengths, but that's because the lengths can be unknown. Here, the lengths are always going to be non-negative integers, and `__len__` is fine to use.